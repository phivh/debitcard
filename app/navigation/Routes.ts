export const Routes = {
  Home: 'Home',
  Debit: 'Debit Card',
  DebitCard: 'DebitCard',
  Payments: 'Payments',
  Credit: 'Credit',
  Profile: 'Profile',
  SpendingLimit: 'SpendingLimit',
};

export default Routes;
