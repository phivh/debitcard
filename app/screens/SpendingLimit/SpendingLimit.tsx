import React, {useCallback, useMemo, useState} from 'react';
import {
  Container,
  Content,
  Currency,
  H2,
  Header,
  Button,
  Text,
  View,
} from 'components';
import {StyleSheet, TouchableOpacity} from 'react-native';
import i18n from 'translations';
import {Colors, Fonts, Spacing} from 'theme';
import {Icon} from 'assets';
import {TextInput} from 'react-native-gesture-handler';
import {convertString2Number, formatCurrency, formatNumber} from 'services';
import {useCard} from 'hooks';

const t = (key: string) => i18n.t(`screens.limit.${key}`);

export const SpendingLimitScreen = () => {
  const {onSetLimit, loading} = useCard();
  const [value, setValue] = useState<number>();
  const onChange = useCallback((v: any) => {
    const number = typeof v === 'object' ? v._value : v;
    setValue(number);
  }, []);
  const limit = useMemo(() => (value ? formatNumber(value) : ''), [value]);
  return (
    <Container testID="spending-limit-screen" loading={loading}>
      <View style={Styles.head}>
        <Header />
        <H2 bold white style={Styles.title}>
          {i18n.t('screens.limit.title')}
        </H2>
      </View>
      <Content bounces={false} style={Styles.content}>
        <View row align="center">
          <Icon name="speedometer-outline" size={20} fill={Colors.black} />
          <Text style={Styles.caption}>{t('caption')}</Text>
        </View>
        <View row style={Styles.field} align="center">
          <Currency />
          <TextInput
            testID="spend-limit-number"
            defaultValue=""
            value={limit}
            keyboardType="decimal-pad"
            onChangeText={(v: string) => onChange(convertString2Number(v))}
            style={Styles.input}
          />
        </View>
        <Text small grey>
          {t('notify')}
        </Text>
        <View style={Styles.pillContainer} row content="between">
          <TouchableOpacity onPress={() => onChange(5000)} style={Styles.pill}>
            <Text center green>
              {formatCurrency(5000)}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => onChange(10000)} style={Styles.pill}>
            <Text center green>
              {formatCurrency(10000)}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity onPress={() => onChange(20000)} style={Styles.pill}>
            <Text center green>
              {formatCurrency(20000)}
            </Text>
          </TouchableOpacity>
        </View>
      </Content>
      <View style={Styles.footer}>
        <Button
          disabled={!value}
          onPress={() => !!value && onSetLimit({limit: value})}
          testID="btn-save"
          title={i18n.t('actions.save')}
          shadow
          style={[Styles.btn, !value && {opacity: 0.5}]}
          titleStyle={Styles.btnText}
        />
      </View>
    </Container>
  );
};

const Styles = StyleSheet.create({
  head: {
    paddingHorizontal: Spacing.default[4],
  },
  title: {
    marginTop: Spacing.unit,
  },
  info: {
    marginTop: Spacing.unit,
  },
  content: {
    backgroundColor: Colors.white,
    flex: 1,
    flexGrow: 1,
    marginTop: Spacing.default[4] * 2,
    height: Spacing.screenHeight,
    borderTopLeftRadius: Spacing.default[4],
    borderTopRightRadius: Spacing.default[4],
    paddingTop: Spacing.default[5],
    paddingHorizontal: Spacing.default[5],
  },
  caption: {
    marginLeft: Spacing.unit,
    letterSpacing: 0.5,
    ...Fonts.medium,
  },
  field: {
    marginTop: Spacing.unit,
    paddingVertical: Spacing.unit,
    borderBottomWidth: 0.5,
    borderBottomColor: Colors.grey,
    marginBottom: Spacing.default[0],
  },
  input: {
    ...Fonts.h2,
    fontWeight: 'bold',
    maxWidth: '80%',
    width: '80%',
  },
  pillContainer: {
    marginTop: Spacing.default[6],
  },
  pill: {
    width: '30%',
    backgroundColor: Colors.greenLight,
    paddingVertical: Spacing.default[1],
    borderRadius: 5,
  },
  footer: {
    backgroundColor: Colors.white,
    paddingBottom: Spacing.default[8],
  },
  btn: {
    backgroundColor: Colors.green,
    width: 300,
    padding: Spacing.padding,
  },
  btnText: {
    textAlign: 'center',
    fontSize: 20,
  },
});
