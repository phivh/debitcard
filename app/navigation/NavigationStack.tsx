import React from 'react';
import {Pressable} from 'react-native';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import {
  BottomTabBarButtonProps,
  createBottomTabNavigator,
} from '@react-navigation/bottom-tabs';

import * as Screens from 'screens';

import Routes from './Routes';
import {navigationRef} from './NavigationService';
import {Colors} from 'theme';
import {Text} from 'components';
import {Icon} from 'assets';

const Stack = createStackNavigator();
const Tab = createBottomTabNavigator();

const DebitStack = () => (
  <Stack.Navigator
    screenOptions={{headerShown: false}}
    initialRouteName={Routes.Debit}>
    <Stack.Screen name={Routes.DebitCard} component={Screens.DebitCardScreen} />
    <Stack.Screen
      name={Routes.SpendingLimit}
      component={Screens.SpendingLimitScreen}
    />
  </Stack.Navigator>
);

const DisabledTabBarButton = ({style, ...props}: BottomTabBarButtonProps) => (
  <Pressable disabled style={[{opacity: 1}, style]} {...props} />
);
export const Navigator = () => (
  <NavigationContainer ref={navigationRef}>
    <Tab.Navigator initialRouteName={Routes.Debit}>
      <Tab.Screen
        name={Routes.Home}
        options={{
          tabBarButton: DisabledTabBarButton,
          tabBarIcon: () => (
            <Icon name="ios-home-outline" size={32} color={Colors.inactive} />
          ),
          tabBarLabel: () => (
            <Text small grey>
              {Routes.Home}
            </Text>
          ),
        }}
        component={Screens.Dummy}
      />
      <Tab.Screen
        name={Routes.Debit}
        component={DebitStack}
        options={() => {
          const currentRoute = navigationRef.current?.getCurrentRoute();
          return {
            tabBarIcon: () => (
              <Icon name="card" size={32} color={Colors.actice} />
            ),
            tabBarLabel: () => (
              <Text small green>
                {Routes.Debit}
              </Text>
            ),
            tabBarVisible: currentRoute?.name !== Routes.SpendingLimit,
          };
        }}
      />
      <Tab.Screen
        name={Routes.Payments}
        options={{
          tabBarButton: DisabledTabBarButton,
          tabBarIcon: () => (
            <Icon name="payment" mIcon size={32} color={Colors.inactive} />
          ),
          tabBarLabel: () => (
            <Text small grey>
              {Routes.Payments}
            </Text>
          ),
        }}
        component={Screens.Dummy}
      />
      <Tab.Screen
        name={Routes.Profile}
        options={{
          tabBarButton: DisabledTabBarButton,
          tabBarIcon: () => (
            <Icon name="person" mIcon size={32} color={Colors.inactive} />
          ),
          tabBarLabel: () => (
            <Text small grey>
              {Routes.Profile}
            </Text>
          ),
        }}
        component={Screens.Dummy}
      />
    </Tab.Navigator>
  </NavigationContainer>
);
