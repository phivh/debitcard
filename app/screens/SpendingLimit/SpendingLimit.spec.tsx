import React from 'react';
import {render} from '@testing-library/react-native';
import {Routes} from 'navigation';

import {SpendingLimitScreen} from './SpendingLimit';
import {NavigatorMock} from '../../../__mocks__/NavigatorMock';

let rendered: any;

const components = [{screen: SpendingLimitScreen, name: Routes.SpendingLimit}];

describe('DebitCard screen', () => {
  it('renders correctly', () => {
    rendered = render(<NavigatorMock components={components} />);
    expect(rendered.toJSON()).toMatchSnapshot();
    expect(rendered.getByTestId('spending-limit-screen')).toBeDefined();
  });

  it('has elements in screen', async () => {
    rendered = render(<NavigatorMock components={components} />);
    expect(rendered.getByTestId('spend-limit-number')).toBeDefined();
    expect(rendered.getByTestId('btn-save')).toBeDefined();
  });
});
