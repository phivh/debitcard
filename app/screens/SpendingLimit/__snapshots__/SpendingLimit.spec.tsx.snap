// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebitCard screen renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenContainer
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RNSScreen
      activityState={2}
      collapsable={false}
      forwardedRef={[Function]}
      nativeID="animatedComponent"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "rgb(242, 242, 242)",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#0D365A",
              "flexGrow": 1,
            }
          }
          testID="spending-limit-screen"
        >
          <View
            style={
              Object {
                "alignItems": undefined,
                "justifyContent": undefined,
                "paddingHorizontal": 20,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderBottomWidth": 0,
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "height": 90,
                  "justifyContent": "space-between",
                  "paddingBottom": 16,
                  "paddingTop": 40,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "marginLeft": -10,
                    "minHeight": 46,
                    "minWidth": 46,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 30,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
              <Image
                source="test-file-stub"
                style={
                  Object {
                    "height": 24,
                    "width": 24,
                  }
                }
              />
            </View>
            <Text
              allowFontScaling={false}
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "AvenirNextLTPro-Bold",
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "marginTop": 8,
                }
              }
            >
              Spending limit
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": undefined,
                "backgroundColor": "#FFFFFF",
                "borderTopLeftRadius": 20,
                "borderTopRightRadius": 20,
                "flex": 1,
                "flexGrow": 1,
                "height": 1334,
                "justifyContent": undefined,
                "marginTop": 40,
                "minHeight": 0,
                "padding": 16,
                "paddingHorizontal": 24,
                "paddingTop": 24,
                "width": undefined,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": undefined,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                allowFontScaling={false}
                style={
                  Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "letterSpacing": 0.5,
                    "marginLeft": 8,
                  }
                }
              >
                Set weekly debit card spending limit
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderBottomColor": "#CCCCCC",
                  "borderBottomWidth": 0.5,
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": undefined,
                  "marginBottom": 12,
                  "marginTop": 8,
                  "paddingVertical": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                    "justifyContent": "flex-start",
                  }
                }
                testID="available-balance"
              >
                <View
                  style={
                    Object {
                      "alignItems": undefined,
                      "backgroundColor": "#03D167",
                      "borderRadius": 4,
                      "justifyContent": undefined,
                      "marginRight": 8,
                      "paddingHorizontal": 8,
                      "paddingVertical": 2,
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontWeight": "bold",
                      }
                    }
                  >
                    S$
                  </Text>
                </View>
              </View>
              <TextInput
                collapsable={false}
                defaultValue=""
                keyboardType="decimal-pad"
                onChangeText={[Function]}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "fontFamily": "AvenirNextLTPro-Bold",
                    "fontSize": 24,
                    "fontWeight": "bold",
                    "maxWidth": "80%",
                    "width": "80%",
                  }
                }
                testID="spend-limit-number"
                value=""
              />
            </View>
            <Text
              allowFontScaling={false}
              style={
                Object {
                  "color": "#CCCCCC",
                  "fontSize": 12,
                  "fontWeight": "normal",
                }
              }
            >
              Here weekly means the last 7 days, not the calendar week
            </Text>
            <View
              style={
                Object {
                  "alignItems": undefined,
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "space-between",
                  "marginTop": 30,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "#EFFCF4",
                    "borderRadius": 5,
                    "opacity": 1,
                    "paddingVertical": 14,
                    "width": "30%",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Object {
                      "color": "#03D167",
                      "fontWeight": "400",
                      "textAlign": "center",
                    }
                  }
                >
                  S$5,000
                </Text>
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "#EFFCF4",
                    "borderRadius": 5,
                    "opacity": 1,
                    "paddingVertical": 14,
                    "width": "30%",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Object {
                      "color": "#03D167",
                      "fontWeight": "400",
                      "textAlign": "center",
                    }
                  }
                >
                  S$10,000
                </Text>
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "#EFFCF4",
                    "borderRadius": 5,
                    "opacity": 1,
                    "paddingVertical": 14,
                    "width": "30%",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Object {
                      "color": "#03D167",
                      "fontWeight": "400",
                      "textAlign": "center",
                    }
                  }
                >
                  S$20,000
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": undefined,
                "backgroundColor": "#FFFFFF",
                "justifyContent": undefined,
                "paddingBottom": 48,
              }
            }
          >
            <View
              accessibilityState={
                Object {
                  "disabled": true,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "#03D167",
                  "borderRadius": 30,
                  "borderWidth": 0,
                  "elevation": 1,
                  "opacity": 0.5,
                  "padding": 16,
                  "shadowColor": "#000000",
                  "shadowOffset": Object {
                    "height": 2,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 4,
                  "width": 300,
                }
              }
              testID="btn-save"
            >
              <Text
                allowFontScaling={false}
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Save
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenContainer>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "rgb(255, 255, 255)",
        "borderTopColor": "rgb(216, 216, 216)",
        "borderTopWidth": 0.5,
        "bottom": 0,
        "elevation": 8,
        "height": 49,
        "left": 0,
        "paddingBottom": 0,
        "paddingHorizontal": 1,
        "position": null,
        "right": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityLabel="SpendingLimit, tab, 1 of 1"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "selected": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
            },
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "flexDirection": "column",
              "justifyContent": "flex-end",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "textAlign": "center",
              },
              Object {
                "color": "rgb(0, 122, 255)",
              },
              Object {
                "fontSize": 10,
              },
              undefined,
            ]
          }
        >
          SpendingLimit
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
